### PR TITLE
 IOperation Node Porting: Non-Leaf Non-Hierarchy Nodes Pt3

### PR DESCRIPTION
**Pull request by 333fred**
This is the next set of non-leaf, non-hierarchy IOperation nodes. With this PR, we're ~55% of the way through the types defined in OperationInterfaces.xml. Once again, commit-by-commit is recommended. You might be able to get away with doing a few commits at a time, but reviewing this all at once is likely to be very hard to do (given that there are 28 ported nodes in this PR).

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @jcouv @jaredpar this is ready for review.

**Comment by 333fred** [comment]
@dotnet/roslyn-ide [`456829a` (#48892)](https://github.com/dotnet/roslyn/pull/48892/commits/456829a7f271db03faa2889e2f959b60d8c2ce07) marks `IArgumentOperation.ParameterSymbol` as nullable, because it can be if the argument is `__arglist`. However, this causes issues with the InlineMethodRefactoring, which assumes that `IArgumentOperation.ParameterSymbol` will never return null. The fix here isn't readily apparent to me: is there a simply way to block `__arglist` in this refactoring so we can assume it's never nullable?

**Comment by 333fred** [comment]
> @dotnet/roslyn-ide [`456829a` (#48892)](https://github.com/dotnet/roslyn/pull/48892/commits/456829a7f271db03faa2889e2f959b60d8c2ce07) marks `IArgumentOperation.ParameterSymbol` as nullable, because it can be if the argument is `__arglist`. However, this causes issues with the InlineMethodRefactoring, which assumes that `IArgumentOperation.ParameterSymbol` will never return null. The fix here isn't readily apparent to me: is there a simply way to block `__arglist` in this refactoring so we can assume it's never nullable?

@CyrusNajmabadi or @jasonmalinowski any pointers?

**Comment by jasonmalinowski** [comment]
Somewhere in https://github.com/dotnet/roslyn/blob/7da3c0f540d3383f8d799d8e30e498a32f1ba455/src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs#L93-L111 maybe?

**Review by jcouv — APPROVED** [review]
Compiler portion LGTM Thanks (iteration 26)

**Comment by 333fred** [comment]
@dotnet/roslyn-ide can I get a review on the IDE portions of this change? For the IDE sections, it's probably easiest to not review commit by commit, and instead just review all of it at once as there are not that many.

File: src/Features/CSharp/Portable/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.Rewriting.cs [hunk 1]
```diff
@@ -60,7 +60,7 @@ private static ExpressionSyntax AsExpressionSyntax(IOperation operation)
             => operation switch
             {
                 IReturnOperation { ReturnedValue: { } value } => (ExpressionSyntax)value.Syntax,
-                IThrowOperation op => ThrowExpression((ExpressionSyntax)op.Exception.Syntax),
+                IThrowOperation { Exception: { } exception } => ThrowExpression((ExpressionSyntax)exception.Syntax),
```
Comment (333fred) [inline src/Features/CSharp/Portable/ConvertIfToSwitch/CSharpConvertIfToSwitchCodeRefactoringProvider.Rewriting.cs:63]: This is checked here: https://github.com/dotnet/roslyn/blob/master/src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs#L113

**Review by CyrusNajmabadi — APPROVED** [review]
IDE side lgtm.

**Comment by cston** [comment]
>                     Interlocked.CompareExchange(ref _lazyChildren, builder.ToImmutableAndFree(), null);

Since `_lazyChildren` is an `IEnumerable<IOperation>`, we're allocating an extra instance for the value, even for the empty case, because of boxing. Consider using `builder.ToArrayAndFree()` instead.

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:5799 in 2942c0f. [](commit_id = 2942c0f426c8d85814ec768dbc1ebd22d960d9aa, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/OperationFactory.cs [hunk 2]
```diff
@@ -15,5 +12,12 @@ public static IInvalidOperation CreateInvalidOperation(SemanticModel semanticMod
         {
             return new InvalidOperation(children, semanticModel, syntax, type: null, constantValue: null, isImplicit: isImplicit);
         }
+
+        public static readonly IConvertibleConversion IdentityConversion = new IdentityConvertibleConversion();
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationFactory.cs:16]: >IdentityConversion [](start = 54, length = 18)

Is this shared instance used in `IOperation` trees? If so, how do we handle the parent pointer?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 3]
```diff
@@ -1987,27 +1974,25 @@ private IExpressionStatementOperation CreateBoundExpressionStatementOperation(Bo
...
-            return CreateTupleOperation(boundConvertedTupleLiteral, boundConvertedTupleLiteral.SourceTuple?.Type, createDeclaration);
-        }
-
-        internal IOperation CreateTupleOperation(BoundTupleExpression boundTupleExpression, TypeSymbol? naturalType, bool createDeclaration)
+        internal IOperation CreateBoundTupleOperation(BoundTupleExpression boundTupleExpression, bool createDeclaration = true)
         {
             SyntaxNode syntax = boundTupleExpression.Syntax;
             bool isImplicit = boundTupleExpression.WasCompilerGenerated;
             ITypeSymbol? type = boundTupleExpression.GetPublicTypeSymbol();
+            TypeSymbol? naturalType = boundTupleExpression switch
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1982]: >TypeSymbol? naturalType [](start = 12, length = 23)

Consider moving this below the `if (syntax is DeclarationExpressionSyntax) { ... }` block.

File: src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs [hunk 4]
```diff
@@ -325,15 +326,17 @@ internal abstract partial class AbstractInlineMethodRefactoringProvider<TMethodD
...
 
                 var parameterToReplaceMap =
                     operationsWithLiteralArgument
                     .Concat(operationsWithIdentifierArgument)
                     .Concat(operationsReadOnlyOnce)
                     .Concat(operationsWithDefaultValue)
                     .ToImmutableDictionary(
-                        keySelector: argument => argument.Parameter,
+                        // We excluded arglist callees, so Parameter will always be non null
+                        keySelector: argument => (argument.Parameter!),
```
Comment (cston) [inline src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs:339]: >( [](start = 49, length = 1)

Minor: Parentheses are not needed.

**Review by cston — APPROVED** [review]
A few comments, none blocking.

**Pull request merged**